### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Nessrine88**
**Result before solving n+1 problem**
2023-11-23_20-36-36 png2
**Result after solving n+1 problem:**
2023-11-23_20-48-50 png3
I have created integration tests using Capybara to thoroughly test each view in our project. The tests are organized to ensure a logical flow and cover various aspects of the user interface.
### User Index Page:
Verifies the visibility of usernames, profile pictures, and the number of posts for each user.
Validates the redirection to a user's show page upon clicking a user.
### User Show Page:

1. Ensures the display of the user's profile picture, username, bio, and the first three posts.
2. Includes a button to view all of a user's posts and verifies the redirection to a user's post index page upon clicking a user's post.

### User Post Index Page:

1. Validates the visibility of the user's profile picture, username, post details, comments, likes, and pagination functionality.
2. Verifies the redirection to a post's show page upon clicking a post.

### Post Show Page:

1. Ensures the visibility of the post's title, author, comments, likes, body, and the details of each comment.
2. These tests provide comprehensive coverage of our application's views and user interactions. Your feedback on this pull request is highly appreciated.

### CODE REVIEW

**Review by FrederickMih — COMMENTED** [review]
Hi Team
While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request :red_circle: 

Let's kindly fix the linter issues. A project cannot be review with failing linter checks.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: spec/integrations/users_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    expect(page).to have_content(user1.name)
+  end
+
+  it 'can see the profile picture of the user' do
+    visit path
+
+    photo = find(:css, '.user-photo')
+
+    expect(photo).to_not be_nil
+  end
```
Comment (FrederickMih) [inline spec/integrations/users_index_spec.rb]: - [x] `user1` was created above with a profile picture. I think we are supposed to test that when the users index page is visited(users_path), the user's profile picture(user1.photo) is visible on the page content

File: spec/integrations/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  before :each do
+    user.save
+    post1.save
+    comment1.save
+    like1.save
+  end
+
+  it 'displays the profile picture for each user' do
+    visit users_path
```
Comment (FrederickMih) [inline spec/integrations/users_show_spec.rb]: - [x] Let's kindly note that we are adding tests for the user show page not index page. Let's write unit test to asserts that we can see the user profile picture(`user`.photo) displayed on this page.

File: spec/integrations/users_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+  it 'can see the bio of the user' do
+    visit "/users/#{user.id}"
+    expect(page).to have_content(user.bio)
+  end
+
+  it 'can see the user\'s  most recent posts' do
+    visit "/users/#{user.id}/posts"
+    expect(page).to have_content(post1.text)
+  end
```
Comment (FrederickMih) [inline spec/integrations/users_show_spec.rb]: - [ ]  Having unit tests help us to keep track of the building blocks. They also help you to check that the refactoring codebase in future won't introduced bugs hence we should be testing exactly how our app is working. To minimized the number of posts in this page, we got a method that retrieved only the user's 3 most recent posts. Let's kindly ensure that if a user created at least 4 posts, only the 3 most recent posts should be seen on this page :+1:

File: spec/integrations/users_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    visit "/users/#{user.id}"
+    expect(page).to have_content('See all posts')
+  end
+
+  it 'clicking an user redirects you to the user show page' do
+    visit "/users/#{user.id}/posts"
+
+    click_link 'See this post'
+    expect(page).to have_current_path("/users/#{post1.author_id}/posts/#{post1.id}")
+  end
```
Comment (FrederickMih) [inline spec/integrations/users_show_spec.rb]: - [x] Hi, there's a test scenario in the requirements that came before this test:
- **Tests that redirects to the post's show page when you click on a user's post.** Please let's kindly add unit test for this aspect.

File: spec/integrations/posts_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    post1.save
+    comment1.save
+    like1.save
+  end
+
+  it 'can see the profile picture of the user' do
+    visit "/users/#{user.id}"
+    photo = find(:css, '.user-photo')
+    expect(photo).to_not be_nil
+  end
```
Comment (FrederickMih) [inline spec/integrations/posts_index_spec.rb]: - [x] As mentioned above, let's kindly refactor this test to make sure that the user's profile picture is displayed on the post index page.

File: spec/integrations/posts_index_spec.rb [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    find(:css, '.user-photo')
+
+    expect(page).to have_content(user.name)
+  end
+
+  it 'can see number of posts of the user' do
+    visit path
+    puts user.posts_counter
+    expect(page).to have_content("Number of posts: #{user.posts_counter}")
+  end
```
Comment (FrederickMih) [inline spec/integrations/posts_index_spec.rb]: - [Optional] We are really supposed to write unit tests that covers the user's workflow and not just a test that will passed at the end :+1:

File: spec/integrations/posts_index_spec.rb [hunk 7]
```diff
@@ -0,0 +1,68 @@
+require 'rails_helper'
+
+describe 'posts index', type: :system do
```
Comment (FrederickMih) [inline spec/integrations/posts_index_spec.rb:3]: - [ ] As stated in the requirements, let's kindly write a unit tests `that asserts you can see a section for pagination if there are more posts than fit on the view.` :+1:

File: spec/integrations/posts_show_spec.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+describe 'posts show', type: :system do
```
Comment (FrederickMih) [inline spec/integrations/posts_show_spec.rb:3]: - [ ] I think one test is left:
- `Tests that asserts you can see the username of each commentor.`

**Review by FrederickMih — CHANGES_REQUESTED** [review]
## Hi :raising_hand: :raising_hand: @Nessrine88 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!





## Highlights :green_circle: 

- All tests suites passed :+1: 
- Descriptive PR title and branch name :+1: 


## Required Changes ♻️
**N/A**
_Check the comments under the review._

## Optional suggestions
**N/A**
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/integrations/users_show_spec.rb [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    post4.save
+    comment1.save
+    comment2.save
+    comment3.save
+  end
+
+  it 'displays the profile picture for each user' do
+    visit users_path(user)
+    expect(page).to have_css('img.user-photo')
+  end
```
Comment (TadesseDev) [inline spec/integrations/users_show_spec.rb]: - [ ] Dear team good job updating your test as suggested by the previous reviewer, but still your test does not assert the appropriate profile picture is visible. Please make sure the `user.photo` or the link for the image `https://as1.ftcdn.net/v2/jpg/00/64/30/06/1000_F_64300613_1TYCvyzq09Go7oEVXkDQypofuaJEwKTo.jpg` is present in the page. 
- [Optional] Also consider updating the description for this test. we are only testing a single user profile picture, hence kindle remove/update the `for each user` part od the `it` blcok.

File: spec/integrations/users_show_spec.rb [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    visit "/users/#{user.id}"
+    expect(page).to have_content(user.bio)
+  end
+
+  it 'displays the user\'s 3 most recent posts' do
+    visit "/users/#{user.id}"
+    expect(page).to have_content(post2.text)
+    expect(page).to have_content(post3.text)
+    expect(page).to have_content(post4.text)
+  end
```
Comment (TadesseDev) [inline spec/integrations/users_show_spec.rb]: - [ ] Dear @Nessrine88, Good job updating your test here to test only 3 posts, but we need to make use of the `recent_posts` method inside the `user_model`, This way we can test [the method itself](https://github.com/Nessrine88/Blog-App/blob/2c21138eabe2c2d8e316fe94bd522ba95b23380a/app/models/user.rb#L17C1-L21C4url) is working fine.

**Review by TadesseDev — CHANGES_REQUESTED** [review]
### Hi Team

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

199733072-bdacca19-8df9-43a7-be93-3aff2f1c15c8

 ## Highlights
 - [x] Good job making most of the changes requested by the previous reviewer. 🙏


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by franclobo — APPROVED** [review]
Hi @Nessrine88, @kazim110 , and @alexansaa, 

# STATUS: APPROVED 😃 

git-merge

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- No linter errors ✔️ 
- All tests passed ✔️ 
- Use Capybara ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**